stl/inc/xmemory

```diff
@@ -781,6 +781,33 @@ struct _Simple_types { // wraps types from allocators with simple addressing for
     using const_pointer   = const value_type*;
 };
 
+#if _HAS_CXX23
+// STRUCT TEMPLATE allocation_result
+template <class _Ptr>
+struct allocation_result {
+    _Ptr ptr;
+    size_t count;
+};
+
+#ifdef __cpp_lib_concepts
+// FUNCTION TEMPLATE allocate_at_least
+template <class _Alloc>
+concept _Has_member_allocate_at_least = requires(_Alloc _Al, size_t _Count) {
+    _Al.allocate_at_least(_Count);
+};

```

Comment:
I know that this is C++23 machinery, but we *could* get away with classic SFINAE and not rely on templates. That said, I am all in favor of using concepts

Also we have an `_Is_allocator_v` trait that we might want to use here.

Finally, we should check whether the return type is correct

Reply:
We shouldn't check the return type. The standard says
> `a.allocate_at_least(n)` if that expression is well-formed

not

> `a.allocate_at_least(n)` if that expression is well-formed and results in an `allocation_result`

so if allocator provides `allocate_at_least` which doesn't return `allocation_result` (or type convertible to it) it shouldn't compile.

Also there is a note:
> Note D: An allocator need not support allocate_at_least, but no default is provided in allocator_traits. If an allocator has an allocate_at_least member, it shall satisify the requirements.

Reply:
Ok you are using it to hard error for invalid allocate_at_least. I like that